### PR TITLE
doc: scripts: Software Maturity Report Generation

### PR DESCRIPTION
**Pull request by gautesl**
Added a script that scans the twister build output and .config
files, and generates a software maturity level report. The
script can be run on a local twister build, but is intended to
be run by the CI and publish the results to Azure.

Also added an initial feature-file containing the desired
features and their respective Kconfig rules.

Signed-off-by: Gaute Svanes Lunde <gaute.lunde@nordicsemi.no>

### CODE REVIEW

**Review by PerMac — COMMENTED** [review]
Edit: copied this to the corresponding part of the code. The scenario I described can happen with twister logic. Twister creates test specification by filling it with args from `common`, but then overwrites it with args defined in the given test. So:
```
common:
  integration_platforms: X
tests:
  test_1:
    integration_platforms: Y
  test_2:
    integration_platforms: Z
```   
will end up with neither configuration using X in the end.

I see one corner case, but I'm not sure if it is worth to complicate the logic to include it:
A common section is overwritten if a test specifies its own integration_platforms. If there is a common section with platform X but all tests have integration_platforms and neither of them contain X -> X won't be used in the end.

**Comment by gautesl** [comment]
> I see one corner case, but I'm not sure if it is worth to complicate the logic to include it: A common section is overwritten if a test specifies its own integration_platforms. If there is a common section with platform X but all tests have integration_platforms and neither of them contain X -> X won't be used in the end.

Where in the code does this happen?

File: doc/_scripts/software_maturity_scanner.py [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+            continue
+
+        # Add common integration platforms
+        if "common" in content and "integration_platforms" in content["common"]:
+            all_boards.update(content["common"]["integration_platforms"])
+
+        # Add each test's integration platforms
+        for test in content["tests"].values():
+            if "integration_platforms" in test:
+                all_boards.update(test["integration_platforms"])
```
Comment (PerMac) [inline doc/_scripts/software_maturity_scanner.py:216]: I see one corner case, but I'm not sure if it is worth to complicate the logic to include it:
A common section is overwritten if a test specifies its own integration_platforms. If there is a common section with platform X but all tests have integration_platforms and neither of them contain X -> X won't be used in the end.

Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py:216]: Ah, yes I see, that is exactly the problem then

Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py:216]: A possible fix:
```suggestion
        # Add common integration platforms
        common_platforms = []
        if "common" in content and "integration_platforms" in content["common"]:
            common_platforms = content["common"]["integration_platforms"]

        # Add each test's integration platforms
        for test in content["tests"].values():
            integration_platforms = common_platforms
            if "integration_platforms" in test:
                integration_platforms = test["integration_platforms"]
            all_boards.update(integration_platforms)
```

**Review by PerMac — APPROVED** [review]

**Comment by carlescufi** [comment]
@jangalda-nsc can you please review this PR?

File: doc/_scripts/software_maturity_scanner.py [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+experimental. If this is an undesired outcome, the feature should be split
+into two separate features, one with the experimental symbols and the other
+with the supported ones.
+
+Copyright (c) 2022 Nordic Semiconductor ASA
+"""
+
+from pathlib import Path
+from typing import Dict, List, Set
+from azure.storage.blob import ContainerClient
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:100]: do we have any file which contains pip requirements for this file?

File: doc/_scripts/software_maturity_scanner.py [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+import kconfiglib
+import os
+import re
+import sys
+import yaml
+import logging
+import argparse
+
+
+AZ_CONN_STR = ";".join(
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:111]: do we want to store connection data in sdk-nrf repo?

File: doc/_scripts/software_maturity_scanner.py [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+
+__version__ = "0.1.0"
+
+logger = logging.getLogger(__name__)
+
+
+def upload_data(filename: str, data: dict, cc: ContainerClient) -> None:
+    """Upload software maturity level report to Azure blob storage.
+
+    Args:
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:141]: in some docstrings args are present and in some not.

File: doc/_scripts/software_maturity_scanner.py [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+    """
+
+    data = yaml.safe_dump(data)
+    bc = cc.get_blob_client(filename)
+
+    try:
+        logger.info("Uploading software maturity level information...")
+        bc.upload_blob(data, overwrite=True)
+        logger.info("Upload complete")
+    except ClientAuthenticationError:
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:154]: shouldn't exceptions be re raised? Without it, we cannot determine if upload succeeded or not, and script's return code will be 0

File: doc/_scripts/software_maturity_scanner.py [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+        A sorted list of all unique SoC names
+    """
+
+    all_boards = set()
+    for sample_path in all_samples:
+        try:
+            with open(sample_path) as sample:
+                content = yaml.safe_load(sample)
+        except (yaml.YAMLError, FileNotFoundError) as err:
+            logger.error(err)
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:206]: again, without checking logs, we are not able to determine if there were errors in script or not

File: doc/_scripts/software_maturity_scanner.py [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+        # Add each test's integration platforms
+        for test in content["tests"].values():
+            integration_platforms = common_platforms
+            if "integration_platforms" in test:
+                integration_platforms = test["integration_platforms"]
+            all_boards.update(integration_platforms)
+
+    # Only use nRF boards, but not nRF51 series
+    all_boards = set(
+        filter(lambda b: b.startswith("nrf") and not b.startswith("nrf51"), all_boards)
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:223]: do we expect that filter might change in time? If so, it will be good to have them stored in global variable

File: doc/_scripts/software_maturity_scanner.py [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+    )
+    all_socs = {b.split("_")[1] for b in all_boards}
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:225]: has every `b` at least 1 `_`? If not it will fail

File: doc/_scripts/software_maturity_scanner.py [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+
+
+def evaluate_rule(vars: Dict[str, kconfiglib.Variable], rule: list) -> bool:
+    """Evaluate a parsed Kconfig rule against a dictionary of Kconfig variables.
+
+    Returns:
+        False if any of the required variables are missing, True otherwise
+    """
+
+    assert rule[0] == "OR"
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:305]: for every assert, please log which rule is not valid

File: doc/_scripts/software_maturity_scanner.py [hunk 10]
```diff
@@ -0,0 +1,554 @@
...
+            if isinstance(item, list):
+                out.update(flatten(item))
+            else:
+                out.add(item)
+        return out
+
+    symbols = flatten(rule)
+    for symbol in symbols:
+        if symbol in experimental_symbols:
+            return "Experimental"
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:344]: you can also use enum/globals to avoid typos

File: doc/_scripts/software_maturity_scanner.py [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
+    return "Supported"
+
+
+def generate_tables(args) -> None:
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:348]: I would prefer to use args only in `__name__ == '__main__'` block. Because, in theory, other methods from script can be run from other python packages. Please change method api to `generate_tables(build_output, input_file, ...)`

File: doc/_scripts/software_maturity_scanner.py [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+
+    # Directory sanity check
+    sample_dir = NRF_SAMPLE_DIR
+    if not args.build_output.exists():
+        sys.exit("Invalid output directory")
+    elif (
+        not (sample_dir / "samples").exists()
+        or not (sample_dir / "applications").exists()
+    ):
+        sys.exit("samples/ or applications/ cannot be found in " + str(sample_dir))
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:363]: please use f-strings

File: doc/_scripts/software_maturity_scanner.py [hunk 13]
```diff
@@ -0,0 +1,554 @@
...
+    # Create a dictionary of all unique features and their respective rule
+    all_features = {}
+    feature_categories = input_doc["features"]
+    for cat, features in feature_categories.items():
+        for feature, rule in features.items():
+            key = f"{cat}_{feature}"
+            all_features[key] = parse_rule(rule)
+
+    # Find all sample.yaml files
+    all_samples = list(sample_dir.glob("samples/**/sample.yaml"))
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:383]: @PerMac do we have any testcase.yamls in samples and applications?

File: doc/_scripts/software_maturity_scanner.py [hunk 14]
```diff
@@ -0,0 +1,554 @@
...
+
+    for feature in all_features:
+        soc_sets[feature] = {}
+
+    # Look for the desired symbol combinations in all .config files
+    all_conf_files = args.build_output.glob("**/zephyr/.config")
+    for conf_file in all_conf_files:
+        kconf = kconfiglib.Kconfig(filename=conf_file)
+        soc = kconf.variables["CONFIG_SOC"].value
+        soc = soc.strip('"').split("_")[0].lower()
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:408]: `soc.strip('"').split("_")[0].lower()` is duplicated in this file, so you can extract it to separate method

File: doc/_scripts/software_maturity_scanner.py [hunk 15]
```diff
@@ -0,0 +1,554 @@
...
+    if args.publish_results:
+        cc = ContainerClient.from_connection_string(
+            AZ_CONN_STR.format(args.access_key), AZ_CONTAINER
+        )
+
+        filename = UPLOAD_FILENAME.format(args.commit_sha)
+        upload_data(filename, output, cc)
+
+    # Save to local file
+    if args.save_output:
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:465]: please save_output first, and then try publishing results (easier to debug if publishing fails)

**Comment by jangalda-nsc** [comment]
I added my comments. From CI perspective, the most important is to determine if script failed or not. So, it needs to return non zero exit code in case of errors. You can decide if error in single file/rule should break CI, but from my experience no one read logs, so it should fail in most cases.

How do we handle this PR
1. Merge it before changing CI and implement needed stages later?
2. Prepare changes in CI before merging this PR, verify it works, then merge PR

I vote on 2 - it will be more efficient if more chnages in script are needed

File: doc/_scripts/software_maturity_scanner.py (hunk 2 above)
Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py:100]: The script seems to work in a fresh virtual environment with python 3.8.10 given these requirements:
```
azure-storage-blob==12.9.0
kconfiglib==14.1.0
PyYAML==5.4.1
```

Should this be stored in a specific file somewhere?

File: doc/_scripts/software_maturity_scanner.py (hunk 3 above)
Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py:111]: This data can already be found in `cache_upload.py` in the same folder

File: doc/_scripts/software_maturity_scanner.py [hunk 16]
```diff
@@ -0,0 +1,554 @@
...
+        "The environment variable TWISTER_OUT_DIR is used by default.",
+    )
+
+    parser.add_argument(
+        "-f",
+        "--input-file",
+        default="software_maturity_features.yaml",
+        type=Path,
+        help="Input YAML file name. "
+        "Defaults to look for 'software_maturity_features.yaml' in the local directory.",
```
Comment (jangalda-nsc) [inline doc/_scripts/software_maturity_scanner.py:491]: I think default should be set to `<script_path>/software_maturity_features.yaml`. Otherwise, script fails if run outsite `docs/_scripts` folder without `--input-file` argument

File: doc/_scripts/software_maturity_scanner.py (hunk 4 above)
Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py]: All docstrings should now be consistent

File: doc/_scripts/software_maturity_scanner.py (hunk 5 above)
Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py]: The exception is no longer catched, preserving the original stack trace. a "Raises" clause has been added to the docstring.

File: doc/_scripts/software_maturity_scanner.py (hunk 6 above)
Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py]: Same solution as above

File: doc/_scripts/software_maturity_scanner.py (hunk 7 above)
Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py]: A `BOARD_FILTER` constant has been added as a global variable, and now uses regex to find "_nrfxxx" instead of relying on string start.

File: doc/_scripts/software_maturity_scanner.py (hunk 8 above)
Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py]: Changed to use regex to find "_(nrf[0-9+])". Guaranteed to match due to the filter above.

File: doc/_scripts/software_maturity_scanner.py (hunk 9 above)
Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py]: The relevant rule is added with an f-string to the asserts

File: doc/_scripts/software_maturity_scanner.py (hunk 10 above)
Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py]: Added a MaturityLevels enum. Only the string values are used in the output however, as it must be compatible with a YAML output format.

File: doc/_scripts/software_maturity_scanner.py (hunk 11 above)
Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py]: All command line arguments are now named arguments in `generate_tables`

File: doc/_scripts/software_maturity_scanner.py (hunk 14 above)
Comment (gautesl) [reply doc/_scripts/software_maturity_scanner.py]: Added a `soc_from_kconfig` function which encompasses this and the previous line

**Review by jangalda-nsc — APPROVED** [review]
